uint8[] data
uint32 width
uint32 height
string format